### PR TITLE
Parsing of GlyphOrderAndAliasDB encoding

### PR DESCRIPTION
**Pull request by felipesanches**
This piece of code implements parsing of GlyphOrderAndAliasDB encoding.
It is equivalent to #2467 but it includes only the set of changes that are not merely cosmetic (such as code cleanup).

### CODE REVIEW

**Comment by JoesCat** [comment]
Thanks Felipe for redoing this.

Some suggestions to look at:
```c
  for (i = sizeof(names)/sizeof(names[0])-1; i >= 0; --i) {
       encs[i] = -1;
       names[i] = NULL;
  }
```
The above looks crazy, but the start point gets computed once vs the end point getting repeatedly computed (for inefficient compilers, but hopefully today's compilers are smarter enough to compute the end-point just once). Also, some processors have branch instructions like 'test' and 'jge', making the end-point a little quicker too). No worries, about your code, but just see it as a speed improvement (especially if dealing with interpretive languages like python).

The next suggestion I suggest as worth revisiting:
```c
  while (fgets(buffer, sizeof(buffer), file) != NULL) {
```
Anywhere you have a file read or write, don't use !=NULL, but preferably check for '<0' because there are other errors that cause problems, for example, file-lock, out of file space, corrupt/unreadable input, etc, etc, etc....
This I recommend revisiting.

...otherwise....comments aside...

Thanks for your help/improvements.
Joe

**Comment by jtanx** [comment]
Hmm. That comment on `fgets` doesn't make sense - [documentation](http://linux.die.net/man/3/fgets) explicitly states it returns either the pointer passed to it (`buffer` in this case), or NULL on error. So checking against NULL is correct.

**Comment by JoesCat** [comment]
On April 19, 2016 08:07:53 AM Jeremy Tan wrote:
> Hmm. That comment on `fgets` doesn't make sense -
> [documentation](http://linux.die.net/man/3/fgets) explicitly states it
> returns either the pointer passed to it (`buffer` in this case), or
> NULL on error. So checking against NULL is correct.

My err, was thinking of fgetc() and fread(), fwrite().
Here's fgets: http://www.drpaulcarter.com/cs/common-c-errors.php#4.4

File: fontforge/encoding.c [hunk 1]
```diff
@@ -576,6 +576,69 @@ return( NULL );
...
+
+    max = -1; any = 0;
+    i = 0;
+    while (fgets(buffer, sizeof(buffer), file) != NULL) {
+        max = i;
+
+        int tab, tab2;
+        for (tab = 0; tab < sizeof(buffer) && buffer[tab] != '\t'; tab++);
+        if (tab < sizeof(buffer))
+            buffer[tab]='\0';
```
Comment (frank-trampe) [inline fontforge/encoding.c]: If the input item is 256 long, it seems that we would feed an unterminated string into UniFromName.

File: fontforge/encoding.c [hunk 2]
```diff
@@ -576,6 +576,69 @@ return( NULL );
 return( item );
 }
 
+/* Parse the GlyphOrderAndAliasDB file format */
+static Encoding *ParseGlyphOrderAndAliasDB(FILE *file) {
+    char* names[1024];
+    char buffer[256];
```
Comment (frank-trampe) [inline fontforge/encoding.c]: I am not familiar with the file format. Is this limit of 255 characters in the specification?

File: fontforge/encoding.c [hunk 3]
```diff
@@ -576,6 +576,69 @@ return( NULL );
...
+
+        if (strcmp(buffer, ".notdef") == 0) {
+            encs[i] = -1;
+        } else if ((enc = UniFromName(buffer, ui_none, &custom)) != -1) {
+            encs[i] = enc;
+
+            /* Used not to do this, but there are several legal names */
+            /*  for some slots and people get unhappy (rightly) if we */
+            /*  use the wrong one */
+            names[i] = copy(&buffer[tab+1]);
```
Comment (frank-trampe) [inline fontforge/encoding.c]: I don't see over-run protection for names.

File: fontforge/encoding.c [hunk 4]
```diff
@@ -576,6 +576,69 @@ return( NULL );
...
+            any = 1;
+        }
+        i++;
+    }
+
+    if (max != -1) {
+        if (++max < 256) max = 256;
+        item = calloc(1,sizeof(Encoding));
+        char* buf = strdup(_("Please name this encoding"));
+        item->enc_name = ff_ask_string(buf, "GlyphOrderAndAliasDB", buf);
```
Comment (frank-trampe) [inline fontforge/encoding.c]: I don't think that ff_ask_string frees the title and message strings.

File: fontforge/encoding.c (hunk 1 above)
Comment (frank-trampe) [inline fontforge/encoding.c]: Wait. Never mind that.

Comment (frank-trampe) [inline fontforge/encoding.c]: But it wouldn't kill us to have some comments.

**Comment by JoesCat** [comment]
Seems I merged this too early.
@felipesanches any fixes? answers?

**Comment by felipesanches** [comment]
I'll review these things right now. Thanks for pointing out.

**Comment by frank-trampe** [comment]
Ping, @felipesanches.

File: fontforge/encoding.c [hunk 5]
```diff
@@ -576,6 +576,69 @@ return( NULL );
...
+            any = 1;
+        } else {
+            names[i] = copy(&buffer[tab+1]);
+            any = 1;
+        }
+        i++;
+    }
+
+    if (max != -1) {
+        if (++max < 256) max = 256;
```
Comment (frank-trampe) [inline fontforge/encoding.c]: Why do we pad the encoding?

**Pull request merged**